### PR TITLE
write an encoder

### PR DESCRIPTION
**Pull request by paultag**
@tianon review pls

### CODE REVIEW

**Comment by paultag** [comment]
@tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon @tianon v

File: control/encode_test.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+`)))
+
+	assert(t, el.Foo == "test")
+
+	writer := bytes.Buffer{}
+	isok(t, control.Marshal(&writer, el))
+	assert(t, writer.String() == `Foo: test
+X-A-Test: Foo
+
+`)
```
Comment (tianon) [inline control/encode_test.go]: Any particular reason this comes with extra whitespace at the end when it's just a single paragraph?

**Comment by tianon** [comment]
Seems pretty sane to me, although the extra newline added to the end is a little strange (I'd expect to add that myself as necessary, especially since most of my own files that implement this spec are single-document, and I wouldn't want the last paragraph to include an extra newline anyhow).

**Comment by paultag** [comment]
@tianon the reason is that this could end up writing a stream of things; perhaps this is a thing I can add to the Unmarshal code.

**Comment by tianon** [comment]
Hmm, yeah -- I guess it is pretty trivial to do `strings.TrimSpace` when you want explicit control of your output. :+1:

**Comment by paultag** [comment]
@tianon I adjusted the API - but I think this might shoot someone in the foot -- basically, the semantics:

if you use a `control.Encoder`, and `.Encode` a few things, it will DTRT.

If you call `control.Marshal` multiple times, it'll fuck you up :)

Thoughts?

**Comment by tianon** [comment]
Hmm, I think that makes sense personally -- I'd see `control.Encoder` as the streaming interface and `control.Marshal` as the one-off interface (maybe even note that in the docs?)

**Comment by paultag** [comment]
@tianon done - LGTY?

**Comment by tianon** [comment]
LGTM :+1:

**Pull request merged**